### PR TITLE
save and restore lr_scheduler as needed

### PR DESCRIPTION
**Pull request by tuglat**
This change fixes the problem in issue #890, where the lr_scheduler does not behave as expected when continuing training. The problem is that the lr_scheduler is kept as part of the optimizer, but the optimizer is not saved when saving state. Therefore, every time training is restarted, a new lr_scheduler is created with initial parameter settings.

#### Pull Request Checklist ##
- [ x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.

- [x] Unit tests pass (`pytest`)
        --- There are failures, but these are the same failures exhibited by the master branch
- [no] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
         --- There are failures, but these are the same failures exhibited by the master branch
- [x] Passed code style checking (`./style-check.sh`)
         --- There are failures, but these are the same failures exhibited by the master branch
- [no] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/lr_scheduler.py [hunk 1]
```diff
@@ -175,8 +176,7 @@ def __call__(self, t):
         return lr
 
     def __repr__(self):
-        return "LearningRateSchedulerPlateauReduce(reduce_factor=%.2f, " \
-               "reduce_num_not_improved=%d)" % (self.reduce_factor, self.reduce_num_not_improved)
+        return "LearningRateSchedulerPlateauReduce(reduce_factor=%.2f, reduce_num_not_improved=%d, num_not_improved=%d, base_lr=%s, lr=%s, warmup=%d, warmed_up=%s)" % (self.reduce_factor, self.reduce_num_not_improved, self.num_not_improved, self.base_lr, self.lr, self.warmup, self.warmed_up)
```
Comment (fhieber) [inline sockeye/lr_scheduler.py:179]: Can you limit line length to 120 here?

File: sockeye/training.py [hunk 2]
```diff
@@ -587,6 +588,18 @@ def _load_trainer_states(self, fname):
         self.trainer.load_states(fname)
         logger.info('Loaded optimizer states from "%s"', fname)
 
+    def _save_lr_scheduler(self, fname):
+        with open(fname, "wb") as fp:
+            pickle.dump(self.trainer.optimizer.lr_scheduler, fp)
+        logger.info('Saved lr_scheduler to "%s"', fname)
```
Comment (fhieber) [inline sockeye/training.py:594]: can we combine / simplify to one line of logging instead?
```suggestion
        logger.info('Saved '%s' to "%s", self.trainer.optimizer.lr_scheduler, fname)
```

File: sockeye/training.py [hunk 3]
```diff
@@ -587,6 +588,18 @@ def _load_trainer_states(self, fname):
...
+        logger.info('Saved lr_scheduler %s', self.trainer.optimizer.lr_scheduler.__repr__())
+
+    def _load_lr_scheduler(self, fname):
+        with open(fname, "rb") as fp:
+            self.trainer.optimizer.lr_scheduler = pickle.load(fp)
+        logger.info('Loaded lr_scheduler from "%s"', fname)
```
Comment (fhieber) [inline sockeye/training.py:600]: same here

File: sockeye/training.py [hunk 4]
```diff
@@ -587,6 +588,18 @@ def _load_trainer_states(self, fname):
         self.trainer.load_states(fname)
         logger.info('Loaded optimizer states from "%s"', fname)
 
+    def _save_lr_scheduler(self, fname):
+        with open(fname, "wb") as fp:
+            pickle.dump(self.trainer.optimizer.lr_scheduler, fp)
+        logger.info('Saved lr_scheduler to "%s"', fname)
+        logger.info('Saved lr_scheduler %s', self.trainer.optimizer.lr_scheduler.__repr__())
+
+    def _load_lr_scheduler(self, fname):
```
Comment (fhieber) [inline sockeye/training.py:597]: can we make this method backwards compatible? Right now, it would be failing when trying to continue training an existing (2.3.5 and lower) model. You could check in the method for existence of the file under `fname` and only load if it exists.

**Review by fhieber — APPROVED** [review]
Thanks for fixing this!

**Review by fhieber — CHANGES_REQUESTED** [review]
Sorry meant to click 'Request Changes'

File: sockeye/lr_scheduler.py (hunk 1 above)
Comment (tuglat) [reply sockeye/lr_scheduler.py:179]: ok

File: sockeye/training.py (hunk 2 above)
Comment (tuglat) [reply sockeye/training.py:594]: ok

File: sockeye/training.py (hunk 3 above)
Comment (tuglat) [reply sockeye/training.py:600]: ok

File: sockeye/training.py (hunk 4 above)
Comment (tuglat) [reply sockeye/training.py:597]: ok

**Review by fhieber — APPROVED** [review]
LGTM, thanks!

**Comment by ketranm** [comment]
LGTM!

File: sockeye/training.py [hunk 5]
```diff
@@ -587,6 +588,17 @@ def _load_trainer_states(self, fname):
         self.trainer.load_states(fname)
         logger.info('Loaded optimizer states from "%s"', fname)
 
+    def _save_lr_scheduler(self, fname):
+        with open(fname, "wb") as fp:
```
Comment (tdomhan) [inline sockeye/training.py:592]: The scheduler can also be `None`. Could you add a check for that?

**Review by tdomhan — CHANGES_REQUESTED** [review]

File: sockeye/training.py (hunk 5 above)
Comment (tuglat) [reply sockeye/training.py]: ok

**Review by tdomhan — APPROVED** [review]
Excellent, thanks for the contribution!
